pennylane_forest/numpy_wavefunction.py

```diff
@@ -100,50 +100,31 @@ def ev(self, A, wires):
           float: expectation value :math:`\left\langle{A}\right\rangle = \left\langle{\psi}\mid A\mid{\psi}\right\rangle`
         """
         # Expand the Hermitian observable over the entire subsystem
-        A = self.expand(A, wires)
-        return np.vdot(self.state, A @ self.state).real
+        As = self.mat_vec_product(A, self.state, wires)
+        return np.vdot(self.state, As).real
 
-    def expand(self, U, wires):
-        r"""Expand a multi-qubit operator into a full system operator.
+    def mat_vec_product(self, mat, vec, wires):
+        r"""Apply multiplication of a matrix to subsystems of the quantum state.
 
         Args:
-            U (array): :math:`2^n \times 2^n` matrix where n = len(wires).
-            wires (Sequence[int]): Target subsystems (order matters! the
-                left-most Hilbert space is at index 0).
+            mat (array): matrix to multiply
+            vec (array): state vector to multiply
+            wires (Sequence[int]): target subsystems
 
         Returns:
-            array: :math:`2^N\times 2^N` matrix. The full system operator.
+            array: output vector after applying ``mat`` to input ``vec`` on specified subsystems
         """
-        if self.num_wires == 1:
-            # total number of wires is 1, simply return the matrix
-            return U
-
-        N = self.num_wires
-        wires = np.asarray(wires)
-
-        if np.any(wires < 0) or np.any(wires >= N) or len(set(wires)) != len(wires):
-            raise ValueError("Invalid target subsystems provided in 'wires' argument.")
-
-        if U.shape != (2 ** len(wires), 2 ** len(wires)):
-            raise ValueError("Matrix parameter must be of size (2**len(wires), 2**len(wires))")
-
-        # generate N qubit basis states via the cartesian product
-        tuples = np.array(list(itertools.product([0, 1], repeat=N)))
-
-        # wires not acted on by the operator
-        inactive_wires = list(set(range(N)) - set(wires))
-
-        # expand U to act on the entire system
-        U = np.kron(U, np.identity(2 ** len(inactive_wires)))
-
-        # move active wires to beginning of the list of wires
-        rearranged_wires = np.array(list(wires) + inactive_wires)
-
-        # convert to computational basis
-        # i.e., converting the list of basis state bit strings into
-        # a list of decimal numbers that correspond to the computational
-        # basis state. For example, [0, 1, 0, 1, 1] = 2^3+2^1+2^0 = 11.
-        perm = np.ravel_multi_index(tuples[:, rearranged_wires].T, [2] * N)
-
-        # permute U to take into account rearranged wires
-        return U[:, perm][perm]
+        mat = np.reshape(mat, [2] * len(wires) * 2)
+        vec = np.reshape(vec, [2] * self.num_wires)
+        axes = (np.arange(len(wires), 2 * len(wires)), wires)
+        tdot = np.tensordot(mat, vec, axes=axes)
```

Comment:
I can't grok these 4 lines, my brain is feeble. Maybe add a comment or two saying what these steps are for exactly? Or rather, why this particular reshaping and identification of axes creates the desired tensor contraction.

Reply:
Added comments :)

Reply:
Awesome!